### PR TITLE
Fix `GeoDataframe` slicing issue by adding `_slice_` method.

### PR DESCRIPTION
**Pull request by thomcom**
## Description

This PR will adds a `_slice` method which is called when a `cudf.DataFrame` is accessed in a variety of ways. It also adds the `name` member to `GeoSeries` that are pulled from the `GeoDataFrame`.

Fixes https://github.com/rapidsai/cuspatial/issues/676

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cuspatial/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

File: python/cuspatial/cuspatial/tests/test_geodataframe.py [hunk 1]
```diff
@@ -241,6 +241,20 @@ def test_post_slice(gpdf, post_slice):
     assert_eq_geo_df(gi[post_slice], cugpdf_back[post_slice])
 
 
+@pytest.mark.parametrize(
+    "inline_slice",
+    [slice(0, 12)]
+    + [slice(0, 10, 1)]
+    + [slice(0, 3, 1)]
+    + [slice(3, 6, 1)]
+    + [slice(6, 9, 1)],
```
Comment (isVoid) [inline python/cuspatial/cuspatial/tests/test_geodataframe.py:250]: This seems like a rather strange way to construct a list? Did black change it? Why not just
```suggestion
    [slice(0, 12),
    slice(0, 10, 1),
    slice(0, 3, 1),
    slice(3, 6, 1),
    slice(6, 9, 1)],
```

File: python/cuspatial/cuspatial/core/geodataframe.py [hunk 2]
```diff
@@ -120,6 +126,35 @@ def _copy_type_metadata(
 
         return type_copied
 
+    def _slice(self: T, arg: slice) -> T:
+        """
+        Overload the _slice functionality from cudf's frame members.
+        """
+        # Collect the Geometry columns and slice them separately
+        columns_mask = cudf.Series(self.columns)
```
Comment (isVoid) [inline python/cuspatial/cuspatial/core/geodataframe.py:134]: cudf does not guarantee performance with wide table. Neither should geodataframe. There's no need to use device to handle the column name list.

File: python/cuspatial/cuspatial/core/geodataframe.py [hunk 3]
```diff
@@ -120,6 +126,35 @@ def _copy_type_metadata(
...
+            self[columns_mask[~geocolumn_mask].values_host]
+        )
+        sliced_data_columns = data_columns._slice(arg)
+        output = {
+            key: (
+                sliced_geo_columns[key] if value else sliced_data_columns[key]
+            )
+            for key, value in zip(
+                columns_mask.values_host, geocolumn_mask.values
+            )
```
Comment (isVoid) [inline python/cuspatial/cuspatial/core/geodataframe.py:154]: I'm a bit not sure if it's name should be called "key" and "value".
```suggestion
            name: (
                sliced_geo_columns[name] if mask else sliced_data_columns[name]
            )
            for name, mask in zip(
                columns_mask.values_host, geocolumn_mask.values
            )
```

File: python/cuspatial/cuspatial/core/geodataframe.py [hunk 4]
```diff
@@ -120,6 +126,35 @@ def _copy_type_metadata(
...
+        }
+        return __class__(output)
```
Comment (isVoid) [inline python/cuspatial/cuspatial/core/geodataframe.py:156]: Interestingly, there's a difference between `__class__` and `self.__class__`. The previous always binds to the current object while the latter gets dynamically dispatched to anything that `self` gets bound to.
```python
In [6]: class A:
   ...:     def __init__(self):
   ...:         print("A ctor")
   ...:     def f(self):
   ...:         print(__class__)
   ...:         print(self.__class__)
   ...: 

In [7]: class B(A):
   ...:     def __init__(self):
   ...:         print("B ctor")
   ...: 

In [8]: b = B()
B ctor

In [9]: b.f()
<class '__main__.A'>
<class '__main__.B'>
```

Comment (isVoid) [reply python/cuspatial/cuspatial/core/geodataframe.py:156]:
```suggestion
        return self.__class__(output)
```

**Review by isVoid — COMMENTED** [review]
I assume this pattern will come up very often: filter out geocolumn, handle geocolumn, handle data column, combine geocolumn and data column. Can we generalize this into some helper functions?

**Comment by thomcom** [comment]
Sure! I have done so in the latest commit. If you approve I'll merge soon.

**Review by isVoid — APPROVED** [review]

**Comment by thomcom** [comment]
@gpucibot merge

**Pull request merged**